File: apps/ml/src/desain_ml/menu_score.py
"""BCG matrix menu scoring. AGENTS.md §28.4 + Glossary.

Categories (Indonesian convention): sapi_perah | bintang | tanda_tanya | anjing.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any


@dataclass(frozen=True)
class ScoreResult:
    menu_item_id: str
    name: str
    category: str
    rationale: str


MIN_PERIOD_DAYS = 30


def score_menu(period_days: int, items: list[Any]) -> list[ScoreResult]:
    if period_days < MIN_PERIOD_DAYS or len(items) == 0:
        return []

    quantities = [it.quantity for it in items]
    margins = [it.margin for it in items]

    median_qty = sorted(quantities)[len(quantities) // 2]
    median_margin = sorted(margins)[len(margins) // 2]

    out: list[ScoreResult] = []
    for it in items:
        high_volume = it.quantity >= median_qty
        high_margin = it.margin >= median_margin

        if high_volume and high_margin:
            cat = "bintang"
            rationale = "Volume tinggi, margin tinggi — pertahankan posisi unggulan."
        elif high_volume and not high_margin:
            cat = "sapi_perah"
            rationale = "Volume tinggi tapi margin tipis — review harga atau resep cost."
        elif not high_volume and high_margin:
            cat = "tanda_tanya"
            rationale = "Margin bagus tapi penjualan rendah — coba dorong via promosi."
        else:
            cat = "anjing"
            rationale = "Volume + margin sama-sama rendah — pertimbangkan dihapus dari menu."

        out.append(
            ScoreResult(
                menu_item_id=it.menu_item_id,
                name=it.name,
                category=cat,
                rationale=rationale,
            )
        )
    return out

File: packages/domain/src/menu-score.ts
/**
 * BCG-matrix classifier for menu items. Used by the menu-score worker and
 * referenced by the proposal §6.2 / §7.6.
 */

export type MenuCategory = 'bintang' | 'sapi_perah' | 'tanda_tanya' | 'anjing';

export type MenuScoreInput = {
  /** Sold units in the period. */
  qty: number;
  /** Gross margin (revenue − cost) in sen. */
  margin: bigint;
  /** Tenant-wide median qty (cutoff for high-volume). */
  qtyCutoff: number;
  /** Tenant-wide median margin (cutoff for high-margin). */
  marginCutoff: bigint;
};

export function classify(input: MenuScoreInput): MenuCategory {
  const highVolume = input.qty >= input.qtyCutoff;
  const highMargin = input.margin >= input.marginCutoff;
  if (highVolume && highMargin) return 'bintang';
  if (highVolume && !highMargin) return 'sapi_perah';
  if (!highVolume && highMargin) return 'tanda_tanya';
  return 'anjing';
}

export function median(arr: number[]): number {
  if (arr.length === 0) return 0;
  const sorted = [...arr].sort((a, b) => a - b);
  const mid = Math.floor(sorted.length / 2);
  return sorted.length % 2 === 0
    ? Math.floor((sorted[mid - 1]! + sorted[mid]!) / 2)
    : sorted[mid]!;
}

export function medianBig(arr: bigint[]): bigint {
  if (arr.length === 0) return BigInt(0);
  const sorted = [...arr].sort((a, b) => (a < b ? -1 : a > b ? 1 : 0));
  const mid = Math.floor(sorted.length / 2);
  if (sorted.length % 2 === 0) {
    return (sorted[mid - 1]! + sorted[mid]!) / BigInt(2);
  }
  return sorted[mid]!;
}

export function rationaleFor(category: MenuCategory, name: string): string {
  switch (category) {
    case 'bintang':
      return `${name} laris dan margin tinggi — pertahankan kualitas, promosikan terus`;
    case 'sapi_perah':
      return `${name} laris tapi margin tipis — coba naikkan harga sedikit atau kurangi food cost`;
    case 'tanda_tanya':
      return `${name} margin bagus tapi kurang laris — pertimbangkan promosi atau reposisi`;
    case 'anjing':
      return `${name} kurang laris dan margin rendah — kandidat untuk dihapus dari menu`;
  }
}
